File: 0107) Binary Tree Level Order Traversal II.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrderBottom(TreeNode* root) {
        vector<vector<int>> ans;
        helper(root, ans, 0);
        reverse(ans.begin(), ans.end());
        return ans;
    }
    void helper(TreeNode *root, vector<vector<int>> &ans, int level)
    {
        if(root == nullptr)
        {
            return;
        }
        if(level == ans.size())
        {
            ans.push_back({});
        }
        ans[level].push_back(root->val);
        helper(root->left, ans, level+1);
        helper(root->right, ans, level+1);
    }
};

File: 0107) Binary Tree Level Order Traversal II.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> levelOrderBottom(TreeNode root) {
        List<List<Integer>> ans = new ArrayList<>();
        int level = 0;
        helper(root, level, ans);
        Collections.reverse(ans);
        return ans;
    }
    public void helper(TreeNode root, int level, List<List<Integer>> ans)
    {
        if(root == null)
        {
            return;
        }
        if(ans.size() == level)
        {
            ans.add(new ArrayList<>());
        }
        ans.get(level).add(root.val);
        helper(root.left, level+1, ans);
        helper(root.right, level+1, ans);
    }
}
